[mlir][bufferization] Don't clone on unknown ownership and verify function boundary ABI

maerhart opened the pull request:
Inserting clones requires a lot of assumptions to hold on the input IR, e.g.,
all writes to a buffer need to dominate all reads. This is not guaranteed by
one-shot bufferization and isn't easy to verify, thus it could quickly lead to
incorrect results that are hard to debug. This commit changes the mechanism of
how an ownership indicator is materialized when there is not already a unique
ownership present. Additionally, we don't create copies of returned memrefs
anymore when we don't have ownership. Instead, we insert assert operations to
make sure we have ownership at runtime, or otherwise report to the user that
correctness could not be guaranteed.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/Transforms/Passes.h:
```diff
@@ -134,8 +135,9 @@ func::FuncOp buildDeallocationLibraryFunction(OpBuilder &builder, Location loc,
-                                              bool privateFuncDynamicOwnership);
+LogicalResult deallocateBuffersOwnershipBased(
+    FunctionOpInterface op,
+    const DeallocationOptions &options = DeallocationOptions());
```
Maybe double check, but I think default values should not be bound to const references.

maerhart replied:
If I understand the standard correctly, this should be fine. The livetime of the temporary object will be extended to that of the const reference bound to. (it would be illegal for non-const references though)

maerhart commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferDeallocationOpInterface.td:
```diff
@@ -56,7 +56,7 @@ def BufferDeallocationOpInterface :
           cannot be easily canonicalized away anymore).
         }],
-        /*retType=*/"std::pair<Value, Value>",
+        /*retType=*/"FailureOr<std::pair<Value, Value>>",
```
I changes the way ownership indicators are materialized in the default case when internal ownership is set to 'Unknown' such that we don't need to return a failure anymore

maerhart commented on mlir/lib/Dialect/Bufferization/IR/BufferDeallocationOpInterface.cpp:
```diff
@@ -132,16 +132,21 @@ void DeallocationState::getLiveMemrefsIn(Block *block,
+    return std::make_pair(memref, ownership.getIndicator());
+
+  if (!options.allowCloning)
+    return emitError(memref.getLoc(),
```
This error does not exist anymore, but I've added testcases for the new ownership indicator materialization mechanism and the insertion of `cf.assert` ops.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferDeallocationOpInterface.h:
```diff
@@ -95,7 +95,17 @@ struct DeallocationOptions {
+
+  /// Inserts `cf.assert` operations to verify the function boundary ABI at
+  /// runtime. Currently, it is only checked that the ownership of returned
+  /// MemRefs is 'true'. This makes sure that ownership is yielded and the
```
`This makes sure that ownership is yielded` -> `This also ensures that the returned memref does not originate from the same allocation as a function argument.`

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferDeallocationOpInterface.td:
```diff
@@ -55,8 +55,38 @@ def BufferDeallocationOpInterface :
+              init_args of `scf.for`)
+            * passing them as additional operands to a call operation when
+              `private-function-dynamic-ownership` is enabled
+            * a copy is made conditionally on the current ownership, etc.
```
do you mean `copy` or `deallocation`?

maerhart replied:
I meant 'copy' when I wrote it, but then removed the cloning, so it's outdated now. Will remove this BP.

matthias-springer commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferDeallocationOpInterface.h:
```diff
@@ -95,7 +95,17 @@ struct DeallocationOptions {
+  /// If it can be determined statically that the ABI is not adhered
+  /// to, an error will already be emitted at compile time. This cannot be
+  /// changed with this option.
+  bool verifyFunctionBoundaryABI = true;
```
I think there are no checks in case of dynamic ownership of private functions (?). That should be mentioned here.

matthias-springer commented on mlir/lib/Dialect/Bufferization/IR/BufferDeallocationOpInterface.cpp:
```diff
@@ -132,30 +132,79 @@ void DeallocationState::getLiveMemrefsIn(Block *block,
+Value DeallocationState::getMemrefWithUniqueOwnership(
+    const DeallocationOptions &options, OpBuilder &builder, Value memref,
+    Block *block) {
+  // NOTE: * if none of the operands have the same allocated pointer, a new
```
indentation is broken

matthias-springer replied:
operands of which op? what does `same` refer to?

matthias-springer commented on mlir/lib/Dialect/Bufferization/IR/BufferDeallocationOpInterface.cpp:
```diff
@@ -132,30 +132,79 @@ void DeallocationState::getLiveMemrefsIn(Block *block,
+  SmallVector<Value> worklist;
+  worklist.push_back(memref);
+
+  while (!worklist.empty()) {
```
What is this worklist loop doing? The deallocation pass already traverses the IR top-to-bottom. Why is it necessary to traverse the use-def chain in reverse here?

maerhart replied:
To only materialize the ownership indicators when needed, but maybe it's better to just eagerly materialize them and rely on DCE to get rid of the ones that are not needed. I also just noticed that the interface implementations that materialize custom ownership indicator logic is not queried here. It would probably also allow us to get rid of the `Ownership` datastructure because 'Unknown' ownerships could not occur anymore (I think).

matthias-springer replied:
That sounds like an interesting idea. (I think that could be a separate NFC pull request?) Can we run into issues where certain unneeded ownership indicators cannot fold away because canonicalizers (for loops, unstructured control flow, etc.) are not efficient enough and the resulting IR is less efficient?

maerhart replied:
In theory that is totally possible and also the reason why I went with this more complicated implementation. E.g., it may be the case that some op with side-effects has to be inserted, just like we create another `arith.select` in its interface implementation but with the difference that this other op is not as pure as `arith.select`. I'm not sure how frequent/relevant such cases are, but I know that in quite a lot of cases we don't care about the ownership of memref values (we only need them for the memrefs we put in the first operand list of the dealloc op, call ops, and ops that forward operands to nested regions (e.g., iter_args of `scf.for`), but not for memrefs we put in the retained list of the dealloc op (there we get the new ownership by other means). In the current implementation, the user of the pass/pipeline can basically choose whether to materialize eagerly (by always calling this materialize method in the `process` method when implementing the interface) or lazily (by default). It's a trade-off between additional pass complexity and providing more flexibility to the user (where I'm not exactly sure if the additional flexibility is actually needed in practice). Maybe just keep it like that and re-evaluate once we know more about the user's needs?

matthias-springer replied:
Yes that makes sense.

maerhart committed:
[mlir][bufferization] Don't clone on unknown ownership and verify function boundary ABI

Inserting clones requires a lot of assumptions to hold on the input IR, e.g.,
all writes to a buffer need to dominate all reads. This is not guaranteed by
one-shot bufferization and isn't easy to verify, thus it could quickly lead to
incorrect results that are hard to debug. This commit changes the mechanism of
how an ownership indicator is materialized when there is not already a unique
ownership present. Additionally, we don't create copies of returned memrefs
anymore when we don't have ownership. Instead, we insert assert operations to
make sure we have ownership at runtime, or otherwise report to the user that
correctness could not be guaranteed.

amrami commented:
Hi!
I have a test case where an argument and a result are aliased.
Today a clone is added and I started looking why.
So I found this commit that was reverted.
Is there a plan to push it? :)
Thanks in advance,
Maya
